[BPK-1376] DialingCodeList and DialingCodeListItem components

shaundon opened the pull request:
# Overview
Adds `BpkDialingCodeList` and `BpkDialingCodeListItem` components.

# Example Usage

## `BpkDialingCodeListItem`
```
<BpkDiallngCodeListItem
    id="44"
    dialingCode="44"
    name="United Kingdom"
    onPress={action('BpkDialingCodeListItem pressed.')}
    flag={
      <Image
        source={{
          uri: 'https://upload.wikimedia.org/wikipedia/commons/thumb/6/6f/Flag_of_the_Democratic_Republic_of_the_Congo.svg/800px-Flag_of_the_Democratic_Republic_of_the_Congo.svg.png',
        }}
      />
    }
/>
```
Note that an instance of `Image` is passed in. I did it this way so that consumers are free to choose whichever type of image source they want: https://facebook.github.io/react-native/docs/images.html.

## `BpkDialingCodeList`
```
const CODES = [
  {
    id: '0',
    dialingCode: '0',
    name: 'Zero',
  },
  {
    id: '1',
    dialingCode: '1',
    name: 'One',
  },
  {
    id: '2',
    dialingCode: '2',
    name: 'Two',
  },
];

<BpkDialingCodeList
  codes={CODES}
  onItemPress={(item) => {
    action(`${item.name} selected`);
  }}
  selectedId="0"
/>
```
* `selectedId` is not required.
* consumers can use `onItemPress` to handle state.

* See storybook for a stateful example.

# Other info

## Android radios
The wireframe for Android looks like this:
![screen shot 2018-03-01 at 15 54 28](https://user-images.githubusercontent.com/73652/36854409-d72c693c-1d68-11e8-8912-8c1f07384cf4.png)

However, React Native doesn't appear to expose these radio buttons (I could be wrong, still making sure). So for now I've used the tick icon, like we have on iOS. We should decide as a squad what to do about this, whether we just add a new icon or what.

## Storybook images
In the storybook, I'm using direct links to Wikipedia for the flags. This probably isn't the best, but I don't love the idea of adding a load of flags to the repo just for this. I'm open to any suggestions!

# GIFs!!

## iOS
![kapture 2018-03-01 at 15 57 46](https://user-images.githubusercontent.com/73652/36854625-5ace0e26-1d69-11e8-9600-af3750a09ea9.gif)

## Android
![kapture 2018-03-01 at 15 58 52](https://user-images.githubusercontent.com/73652/36854671-78cb517c-1d69-11e8-9d1b-51629a1ca2c4.gif)

# To do
* [x] Boilerplate
* [x] Rendering properly
* [x] Selection rendering
* [x] Render tick (iOS)
* [x] Render radios (Android)
* [x] Flag support
* [x] Flags in storybook example
* [x] Tests
* [x] Changelog
* [x] Fix that one Flow error I can't get rid of!
* [x] RTL
* [x] a11y
* [x] Rebase
* [x] Test on different screen sizes

k0nserv commented:
A few thoughts:

+ Should we expose `BpkDiallingCodeListItem` at all? My thoughts are no.
+ Should the dialing code not include the `+`?
+ Should we use something more appropriate for lookups as the structure. Maybe
```javsacript
type Id = string;
type CountryData = {| dialingCode: string, name: string, id: Id |};
type Data = { [key: Id]: CountryData };
```

k0nserv commented on native/packages/react-native-bpk-component-phone-input/src/BpkDiallingCodeList.ios.js:
```diff
@@ -0,0 +1,147 @@
+  b: [{name: 'Bar'}, {name: 'Baz'}],
+}
+*/
+const createAlphabetisedList = data => {
```
This can be simplified as

```javascript
_.groupBy(data, x => x.name[0].toLowerCase());
```

If we are okay with using lodash

shaundon replied:
Good tip, thanks :)

shaundon commented:
> Should we expose BpkDiallingCodeListItem at all? My thoughts are no.

Yeah perhaps not. Reason I did it here was because I matched how `BpkStarRating` exposes `BpkStar`, but we don't have to match that.

> Should the dialing code not include the +?

Could go either way - I decided to have the component add it in, but very happy to change it.

> Should we use something more appropriate for lookups as the structure?

Maybe - interested to see what others think here. So you're saying from a consumer point of view, they'd pass in data like this?:

```
{
  '44': {dialingCode: '44', name: 'United Kingdom', id: '44' }
}
```

Your spelling of `dialing` also made me look up whether it should be `dialing` or `dialling` - looks like U.S. English, which our codes adheres to, only uses a single `L`. So I'll update the code.

k0nserv commented on native/packages/react-native-bpk-component-phone-input/src/BpkDiallingCodeList.android.js:
```diff
@@ -0,0 +1,77 @@
+};
+
+const BpkDiallingCodeList = ({ codes, selectedId, onItemPress }: Props) => {
+  codes.sort((a, b) => {
```
This seems redudant to do in every render, can't we just delegate ordering to consumers?

shaundon replied:
Yeah let's just do that - I'll remove it.

shaundon replied:
They'll get the side-effect that way of being able to use different orders if they want too (e.g. show the current country at the top or something).

k0nserv commented on native/packages/react-native-bpk-component-phone-input/src/BpkDiallingCodeList.ios.js:
```diff
@@ -0,0 +1,147 @@
+  LIST_COMMON_DEFAULT_PROPS,
+} from './common-types';
+
+const spacing = spacingMd * 1.5;
```
This should be spacingBase(16pt) and maybe it could be given a more descriptive name?

shaundon replied:
I've changed `spacingMd * 1.5` to `spacingBase` across the board and it looks too spaced out now:

<img width="244" alt="screen shot 2018-03-02 at 08 48 56" src="https://user-images.githubusercontent.com/73652/36890637-90aa1b56-1df6-11e8-8857-9bec06bb0249.png">

I should also point out that when I spoke to James about this before he left, he said he'd just used values from iOS and not Backpack tokens, which is why we agreed to hack it together with `spacingMd * 1.5`. In Sketch, the vertical padding is 12, hence the calculation.

What do you reckon?

k0nserv replied:
Yes agree we should use the iOS defaults, but the iOS default is 16pt AFAIK. That's also what I am seeing in Sketch with the UI library builtin to 49

k0nserv replied:
Also just to be clear I'm only talking about the leading margin being 16pt here not the vertical padding of the cell. Cells should be 44pt high if we align with the default

shaundon replied:
Ah yeah I see it a lot more now I'm testing on device. I've updated it accordingly!

k0nserv commented on native/packages/react-native-bpk-component-phone-input/src/BpkDiallingCodeList.ios.js:
```diff
@@ -0,0 +1,147 @@
+
+const BpkDiallingCodeList = ({ codes, selectedId, onItemPress }: Props) => (
+  <SectionList
+    sections={convertCodesIntoSections(codes)}
```
I'd wanna look at the affects of this in render and a long list of say 196 entries.

shaundon replied:
I've made a comment further down about performance.

k0nserv commented on native/packages/react-native-bpk-component-phone-input/src/BpkDiallingCodeListItem.ios.js:
```diff
@@ -0,0 +1,128 @@
+} from 'bpk-tokens/tokens/base.react.native';
+import { type ListItemProps, LIST_ITEM_PROP_TYPES } from './common-types';
+
+const padding = spacingMd * 1.5;
```
`spacingBase` here too

k0nserv commented on native/packages/react-native-bpk-component-phone-input/src/common-types.js:
```diff
@@ -0,0 +1,52 @@
+import PropTypes from 'prop-types';
+
+export type ListItemProps = {
+  id: string,
```
Create a new type for `Id`.

```javascript
export type Id = string;
```

k0nserv commented:
> Maybe - interested to see what others think here. So you're saying from a consumer point of view, they'd pass in data like this?:
>```
>{
>  '44': {dialingCode: '44', name: 'United Kingdom', id: '44' }
>}

Yes

matteo-hertel commented on native/packages/react-native-bpk-component-phone-input/index.js:
```diff
@@ -0,0 +1,22 @@
+
+import BpkDialingCodeList from './src/BpkDialingCodeList';
+import BpkDialingCodeListItem from './src/BpkDialingCodeListItem';
+
```
we should expose the PropsTypes here see https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-button/index.js#L23

matthewdavidson replied:
The flow types you mean?

shaundon replied:
Have now done so 👍

matteo-hertel commented on native/packages/react-native-bpk-component-phone-input/package.json:
```diff
@@ -0,0 +1,23 @@
+  "author": "Backpack Design System <backpack@skyscanner.net>",
+  "license": "Apache-2.0",
+  "peerDependencies": {
+    "react": "^16.0.0-alpha.1",
```
both react and native should be the same as the main package.json https://github.com/Skyscanner/backpack/blob/master/native/package.json#L22

k0nserv replied:
I believe we need to add a card to update these across the board

shaundon replied:
Ok cool, I'll leave this for then.

matteo-hertel replied:
Do we have a card or do we need to create one?

matteo-hertel commented on native/packages/react-native-bpk-component-phone-input/src/BpkDialingCodeList.android.js:
```diff
@@ -0,0 +1,77 @@
+};
+
+const BpkDialingCodeList = ({ codes, selectedId, onItemPress }: Props) => {
+  codes.sort((a, b) => {
```
If you move the sort function implementation outside the stateless component IMHO it'll make more readable and maintainable (feel free to disagree)

shaundon replied:
That's a good idea, and I *would* do it, but Hugo's suggested we leave sorting to consumers so I'm just gonna remove it 💅

matteo-hertel replied:
Aight, in which case let's make sure we add that to the docs / readme

matteo-hertel commented on native/storybook/storybook.js:
```diff
@@ -56,7 +56,7 @@ const styles = StyleSheet.create({
     flexDirection: 'column',
     justifyContent: 'center',
-    padding: spacingBase,
+    // padding: spacingBase,
```
this will affect all the stories, is that what we want?

shaundon replied:
Definitely not! I committed that by mistake 😅

shaundon commented:
With regards to performance with long lists, I added a load of countries to the array and the performance on device was shocking (iPhone 7 Plus and Pixel 2). Scrolling was more or less ok, but changing the selected item was taking seconds.

After a bit of googling I came across the solution of using rolling your own `shouldComponentUpdate`, so now the list item components have this:

```
  shouldComponentUpdate(nextProps: Props) {
    return nextProps.selected !== this.props.selected;
  }
```

This sped things up massively. There is still a slight lag when there are >150 items but it's IMO not too bad.

cc @k0nserv @matteo-hertel

shaundon commented:
Update:
* Consumers now need to add the `+`.
* Updated storybook and readme to use ISO country codes for ids.
* Fixed layout issues.
* Confirmed it works on small screen sizes (tested on iPhone SE).
* Set `private: true` in `package,json`.

@k0nserv thanks for the approval, I'll merge once CI passes. I still need to make the Android radio buttons, I'll do that post-merge.

matteo-hertel commented on native/ios/native.xcodeproj/project.pbxproj:
```diff
@@ -5,6 +5,7 @@
 	};
 	objectVersion = 46;
 	objects = {
+
```
How come this file changed? are there new native deps?

shaundon replied:
No clue lol
